File: src/TinRoll.Logic/Manager/QuestionManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Text;
using System.Threading.Tasks;
using TinRoll.Data.Repository;
using TinRoll.Logic.Manager.Interface;
using TinRoll.Logic.Mapper;
using TinRoll.Shared;

namespace TinRoll.Logic.Manager
{
    public class QuestionManager : IQuestionManager
    {
        private readonly IQuestionRepository _questionRepo;

        public QuestionManager(IQuestionRepository questionRepo)
        {
            _questionRepo = questionRepo;
        }

        public async Task<QuestionDto> CreateQuestionAsync(QuestionDto question)
        {
            var dbQuestion = QuestionMapper.ToDb(question);
            var createdQuestion = _questionRepo.CreateQuestion(dbQuestion);
            return QuestionMapper.ToDto(createdQuestion);
        }

        public async Task<QuestionDto> GetQuestionAsync(int id)
        {
            var dbQuestion = _questionRepo.GetQuestion(id);
            return QuestionMapper.ToDto(dbQuestion);
        }

        public async Task<IEnumerable<QuestionDto>> GetQuestionsAsync()
        {
            var dbQuestions = _questionRepo.GetQuestions();
            var questions = dbQuestions.Select(q => QuestionMapper.ToDto(q));
            return questions;
        }
    }
}

File: src/TinRoll.Logic/Manager/QuestionManager.fs
namespace TinRoll.Logic.Manager

open TinRoll.Shared
open TinRoll.Data.Repository
open TinRoll.Logic.Mapper
open System.Threading.Tasks


type IQuestionManager =
    interface
        abstract member GetQuestionAsync : int -> Task<QuestionDto>
        abstract member GetQuestionsAsync : unit -> Task<QuestionDto list>
        abstract member CreateQuestionAsync : QuestionDto -> int Task
    end

type QuestionManager(questionRepo: IQuestionRepository) =
    interface IQuestionManager with
        member this.GetQuestionAsync id =
            let asyncOp = async {
                let dbQuestion = questionRepo.GetQuestionAsync id
                                 |> Async.RunSynchronously
                return match dbQuestion with
                       | Some q -> QuestionMapper.ToDto q
                       | None -> Unchecked.defaultof<QuestionDto>
            }
            Async.StartAsTask(asyncOp)

        member this.GetQuestionsAsync () =
            let asyncOp = async {
                return questionRepo.GetQuestionsAsync ()
                |> Async.RunSynchronously
                |> List.map QuestionMapper.ToDto
            }
            Async.StartAsTask(asyncOp)

        member this.CreateQuestionAsync question =
            let dbQuestion = QuestionMapper.ToDb question
            let asyncOp = async {
                return questionRepo.CreateQuestionAsync dbQuestion
                       |> Async.RunSynchronously
            }
            Async.StartAsTask(asyncOp)
            //let dbQuestion = QuestionMapper.ToDb question
            //questionRepo.CreateQuestionAsync dbQuestion
            //    |> Async.RunSynchronously